src/IW/Effects/Download.hs

```diff
@@ -41,4 +41,4 @@ downloadFileImpl Url{..} = do
             pure $ toStrict body
         _   -> do
             log W $ "Couldn't download file from " <> unUrl
-            throwError notFound
+            pure ""
```

Comment:
This is actually a bit unpleasant moment. I think that if a file is failed to download, this should be an error, not just empty string with logging. And then later in the `getCabalCategories` you can handle the error appropriately using `catchError` function. I would add one more constructor `UrlDownloadFailed Url` to the `AppErrorType` data type to have a descriptive name of this error instead of generic `notFound`. But the general idea is that the `MonadDownload` effect can be used not only by `MonadCabal` and it shouldn't know about the implementation details of classes that are going to use it. Not being able to download a fiel is a valid error.

Reply:
@chshersh got it. I forgot to ask about this.

Reply:
@chshersh 
```
getCabalCategoriesImpl
    :: forall env m.
       WithCabal env m
    => RepoOwner
    -> RepoName
    -> m [Category]
getCabalCategoriesImpl repoOwner repoName = do
    cabalFile <- downloadFile cabalUrl `catchError` urlDownloadFailedHandler
    case parseGenericPackageDescriptionMaybe cabalFile of
        Nothing -> do
            log W $ "Couldn't parse file downloaded from " <> unUrl cabalUrl
            pure []
        Just genPkgDescr -> do
            log I $ "Successfully parsed file downloaded from " <> unUrl cabalUrl
            pure $ categoryNames genPkgDescr
  where
    cabalUrl :: Url
    cabalUrl = repoCabalUrl repoOwner repoName

    urlDownloadFailedHandler :: AppErrorType -> m ByteString
    urlDownloadFailedHandler _ = pure ""
```
Is this fine for handling the error? Or is there a better way? I don't like the fact that the handler ignores the `AppErrorType`, but at the same time pattern matching on each error seems like excessive when it doesn't matter. Maybe the `catchError` should be later in the `do` block? After the part where the `cabalFile` is parsed.

Reply:
This works as well and looks better imo:
```
getCabalCategoriesImpl
    :: forall env m.
       WithCabal env m
    => RepoOwner
    -> RepoName
    -> m [Category]
getCabalCategoriesImpl repoOwner repoName = do
    { cabalFile <- downloadFile cabalUrl;
      case parseGenericPackageDescriptionMaybe cabalFile of
          Nothing -> do
              log W $ "Couldn't parse file downloaded from " <> unUrl cabalUrl
              pure []
          Just genPkgDescr -> do
              log I $ "Successfully parsed file downloaded from " <> unUrl cabalUrl
              pure $ categoryNames genPkgDescr
    } `catchError` handler
  where
    cabalUrl :: Url
    cabalUrl = repoCabalUrl repoOwner repoName

    handler :: AppErrorType -> m [Category]
    handler _ = pure []
```

Reply:
@rashadg1030 `do` with curly braces is weird... Let's leave this syntax to GHC developers 🙂 

I indeed like the second option better because returning `empty` list rather than trying to parse empty text is a saner approach. However, your `catchError` is attached to the whole function body where in reality we are interested in catching error only for `downloadFile` function. It doesn't affect behavior in this case (though it can in more general case) but it makes code more readable.

If I had to write this function, I would implement it like this. First, introduce `downloadFileMaybe` wrapper around `downloadFile`:

```haskell
downloadFileMaybe :: (MonadDownload m, WithError m) => Url -> m (Maybe ByteString)
downloadFileMaybe url = downloadFile url `catchError` \case
    UrlDownloadFailed _ -> pure Nothing
    err -> throwError err
```

And then you can use this function:

```haskell
getCabalCategoriesImpl repoOwner repoName = do
    maybeCabalFile <- downloadFileMaybe cabalUrl
    case maybeCabalFile >>= parseGenericPackageDescriptionMaybe of
        ...
```

(didn't compile, so there might be some mistakes, but I hope that idea is more or less clear)

Reply:
@chshersh This is clear👍🏽 Thank you.

Reply:
@chshersh It won't compile becasue the type of `catchError` doesn't allow it. 
```
-- | Specialized version of 'E.catchError'.
catchError :: WithError m => m a -> (AppErrorType -> m a) -> m a
catchError action handler = action `E.catchError` (handler . appErrorType)
{-# INLINE catchError #-}

downloadFileMaybe :: MonadDownload m => Url -> m (Maybe ByteString)
downloadFileMaybe url = downloadFile url `catchError` \case
    UrlDownloadFailed _ -> pure Nothing
    err -> throwError err
```
`downloadFile` would have to return a `m (Maybe ByteString)`, but it just returns a `m ByteString`. Should the `downloadFileImpl` be modified to return the correct type?

Reply:
I also tried to use something like this, but didn't work out

```
-- | Specialized version of 'E.catchError'.
catchErrorMaybe :: WithError m => m a -> (AppErrorType -> m (Maybe a)) -> m (Maybe a)
catchErrorMaybe action handler = action `E.catchError` (handler . appErrorType)
{-# INLINE catchErrorMaybe #-}
```

Reply:
@rashadg1030 This is a common case. Try this code:

```haskell
(Just <$> downloadFile url) `catchError` \case
```

Functors for the win

Reply:
@chshersh Bingo! Thank you.